### ISSUE TITLE
Does RunningCommand.__eq__ make sense?

### ISSUE THREAD

**Issue by amoffat** [amoffat/sh#518]
It was the root cause of https://github.com/amoffat/sh/issues/413

Here is [`RunningCommand.__eq__`](https://github.com/amoffat/sh/blob/develop/sh.py#L923). Basically the issue was that attempting to call `list.remove(p)` on a list of processes was forcing many of those processes to evaluate/finish, due to `__eq__` indirectly calling `wait()`.

The question is, does it even make sense to compare processes like this? You might be able to have `__eq__` just compare pids, but pids can also be re-used so it isn't perfect.

@ecederstrand what do you think?

**Comment by ecederstrand** [comment]
I'm not sure if the code was changed since, but I don't see how `__eq__` indirectly calls `wait()`. The link does not point to `RunningCommand.__eq__` anymore.

I don't know of any sane way in Unix of knowing if two possibly-running processes are in fact the same. As you say, a pid may be recycled, although I think it's unlikely to happen with the number of processes expected to be in any data structure in `sh`.

Calling `list.remove(p)` on a data structure of a size where you are worried about pids being reused hints to me that you should either redesign to not have that list, or use a collection that supports fast removal, e.g. `set()`. Maybe it should be on the shoulders of the caller to only use `RunningCommand` in a way where equality via pids is safe.

Anyway, I think the expectation of any Python programmer is that `__eq__` is fast. Waiting for a process to exit goes against that expectation.

**Comment by ecederstrand** [comment]
Oh, now I see it. `RunningCommand.__eq__` calls `__unicode__` which accesses `RunningCommand.stdout` which calls `RunningCommand.wait()`.

I also think the expectation is that `__str__` and `__unicode__` are fast. If the process is still running, I think those two methods should detect that and use a different code path that avoids waiting for process completion.

**Comment by amoffat** [comment]
I think I want to keep the behavior of `__str__` and `__unicode__` evaluating the process, but I'm open to changing it for a 2.0 version. My reasoning has always been that if subprocesses were indeed called like functions, then the return value of that function would be a string (the output of the process). Obviously this doesn't work in all cases, like cases where you actually need a `RunningCommand` object to `.wait()` on later. But this is why `RunningCommand` has always behaved very much like a string first, and a `RunningCommand` second. It also has the benefit of being `__repr__`'d in the terminal:

```
>>> import sh
>>> sh.pwd()
/home/amoffat
```

So unless there is a better way to preserve the ideal behavior of a process-function returning string-like output, I think sticking with `__unicode__` evaluating the process is desired for the majority of people. We should open a separate issue on this if you or anyone else feels strongly about it though.

>Calling list.remove(p) on a data structure of a size where you are worried about pids being reused hints to me that you should either redesign to not have that list, or use a collection that supports fast removal, e.g. set(). Maybe it should be on the shoulders of the caller to only use RunningCommand in a way where equality via pids is safe.

There is also the option of just relying on the `RunningCommand` object identity, like Python does by default for object instances. I think the only benefit to using the pid as the `__eq__` is if you could reconstruct the `RunningCommand` object later, from an existing pid id, and use that newly-constructed object to remove another `RunningCommand` object from a collection. This isn't possible currently, and likely won't be possible, so maybe using Python's default `id(ob)` is fine.

**Comment by ecederstrand** [comment]
Thanks for the explanation! In that case, I agree that `id(ob)` is a better fit for `__eq__`. Also, comparing commands by command output seems a bit weird:

```python
>>> import sh
>>> rc1 = sh.echo('foo')
>>> rc2 = sh.cat('foo.txt')  # File created as: echo 'foo' > foo.txt
>>> rc1
foo

>>> rc2
foo

>>> rc1 == rc2
True
```

**Comment by amoffat** [comment]
Landed in 2.0.0
